src/shared/networkrequest.cpp

```diff
@@ -0,0 +1,487 @@
+    }
+  }
+#endif
+
+  int status = statusCode();
+
+  QString expect =
+      m_expectedStatusCode ? QString::number(m_expectedStatusCode) : "any";
+  logger.debug() << "Network reply received - status:" << status
+                 << "- expected:" << expect;
+
+  QByteArray data = m_reply->readAll();
+  processData(m_reply->error(), m_reply->errorString(), status, data);
+}
+
+void NetworkRequest::processData(QNetworkReply::NetworkError error,
+                                 const QString& errorString, int status,
+                                 const QByteArray& data) {
+  m_completed = true;
+  m_timer.stop();
+
+#ifdef MZ_WASM
+  m_finalStatusCode = status;
+#endif
+
+  if (error != QNetworkReply::NoError) {
+    QUrl::FormattingOptions options = QUrl::RemoveQuery | QUrl::RemoveUserInfo;
+    logger.error() << "Network error:" << errorString
+                   << "status code:" << status
+                   << "- body:" << logger.sensitive(data);
+    logger.error() << "Failed to access:" << m_request.url().toString(options);
+    emit requestFailed(error, data);
+    return;
+  }
+
+  // This is an extra check for succeeded requests (status code 200 vs 201, for
+  // instance). The real network status check is done in the previous if-stmt.
+  if (m_expectedStatusCode && status != m_expectedStatusCode) {
+    logger.error() << "Status code unexpected - status code:" << status
+                   << "- expected:" << m_expectedStatusCode;
+    emit requestFailed(QNetworkReply::ConnectionRefusedError, data);
+    return;
+  }
+
+  emit requestCompleted(data);
+}
+
+bool NetworkRequest::isRedirect() const {
+  int status = statusCode();
+  return status >= 300 && status < 400;
+}
+
+void NetworkRequest::handleHeaderReceived() {
+  // Suppress this signal if a redirect is about to happen.
+  int policy =
+      m_request.attribute(QNetworkRequest::RedirectPolicyAttribute).toInt();
+  if (isRedirect() && (policy != QNetworkRequest::ManualRedirectPolicy)) {
+    return;
+  }
+
+  logger.debug() << "Network header received";
+  emit requestHeaderReceived(this);
+}
+
+void NetworkRequest::handleRedirect(const QUrl& redirectUrl) {
+#if QT_VERSION >= 0x060000 && QT_VERSION < 0x060400
+  if (redirectUrl.host().isEmpty()) {
+#  ifdef MZ_DEBUG
+    // See https://bugreports.qt.io/browse/QTBUG-100651
+    logger.debug()
+        << "QT6 redirect bug! The redirected URL is broken because it's not "
+           "resolved using the previous HTTP redirection as base-URL";
+    logger.debug() << "Broken URL:" << redirectUrl.toString();
+    logger.debug() << "Latest redirected URL:" << m_redirectedUrl.toString();
+#  endif
+
+    if (m_redirectedUrl.isEmpty()) {
+      m_redirectedUrl = url().resolved(redirectUrl);
+    } else {
+      m_redirectedUrl = m_redirectedUrl.resolved(redirectUrl);
+    }
+  } else {
+    m_redirectedUrl = redirectUrl;
+  }
+  emit requestRedirected(this, m_redirectedUrl);
+#else
+  emit requestRedirected(this, redirectUrl);
```

Comment:
Qt has a `redirected` signal, curious why we need this.

Reply:
Because we don't want to expose our internal QNetworkReply. The first argument (or the sender()) is the NetworkRequest object.